### DISCUSSION TITLE
Projects using Extism in the wild?

### DISCUSSION THREAD

**Discussion by aschrijver** [extism/extism#684] [General]
In this thread I would like anyone to drop known open source projects that use Extism already. I couldn't easily find examples, but if you know of a project, then don't hesitate and drop a link below :)

**Comment by aschrijver** [comment]
## [Moonrepo WASM plugins](https://moonrepo.dev/docs/guides/wasm-plugins)

> Our WASM plugin system is powered by [Extism](https://extism.org/), a Rust-based cross-language framework for building WASM plugins under a unified guest and host API. Under the hood, Extism uses [wasmtime](https://wasmtime.dev/) as its WASM runtime.
>
>For the most part, you do not need to know about Extism's host SDK, as we have implemented the bulk of it within moon and proto directly. However, you should be familiar with the guest PDKs, as this is what you'll be using to implement Rust-based plugins.

**Reply by nilslice** [reply]
Thanks, @milesj!

**Comment by neuronicnobody** [comment]
## [Simplism](https://github.com/bots-garden/simplism)

> Simplism is a tiny HTTP server to serve [Extism](https://extism.org/) WebAssembly plug-ins and execute/call a single WebAssembly function.

**Reply by nilslice** [reply]
Thanks, @k33g!

**Comment by neuronicnobody** [comment]
## [Otorishi WASM plugins](https://maif.github.io/otoroshi/manual/plugins/index.html)

**Reply by nilslice** [reply]
Thanks, @Zwiterrion!

**Comment by neuronicnobody** [comment]
## [Reactables](https://github.com/dylibso/reactables)

> Reactables bring the power of React and JSX to everyone's programming language of choice. Compile and render JSX templates securely from over 15 different programming languages including Python, Go, Java, Ruby, .NET, PHP, and more!

> Reactables is distributed as an [Extism Plug-in](https://extism.org/docs/concepts/plug-in), so you can now compile and render React/JSX safely from within your programming language of choice.

**Comment by nilslice** [comment]
## [Matricks](https://github.com/wymcg/matricks)

> A command-line tool for extensible LED matrix control with Raspberry Pi devices.

> Matricks is a WASM-based extensible LED matrix control tool intended for use on Raspberry Pi devices. LED matrix functionality is defined by user-developed plugins, or "tricks", which can be developed in any language that is supported by the [Extism PDK](https://extism.org/docs/category/write-a-plug-in). To simulate plugins while you're developing them, check out [Simtricks](https://github.com/wymcg/simtricks)!

**Reply by nilslice** [reply]
https://wymcg.github.io/matricks/

**Reply by nilslice** [reply]
An awesome simulator for the LED panel: https://github.com/wymcg/simtricks

I love how this demonstrates the portability of Wasm so that you can develop your plugin, and test it without _any_ of the hardware. Then run it exactly the same on the hardware device

**Reply by nilslice** [reply]
Thanks, @wymcg!

**Comment by nilslice** [comment]
## [fractal-trick](https://github.com/TheCatster/fractal-trick)

> A Fractal plugin for Matricks

**Reply by nilslice** [reply]
Thanks, @TheCatster!

**Comment by nilslice** [comment]
## [SSD - Simple Service & Data Description](https://github.com/ssd-codegen/ssd)

> What is SSD?
> First and foremost it's supposed to be data format for describing data structures and services. Additionally it provides tooling to work with the aforementioned format, which simplifies writing custom code generators.

https://ssd-codegen.github.io/ssd/

**Reply by nilslice** [reply]
Thanks, @hardliner66 & @ModProg!

**Comment by nilslice** [comment]
## [Image Processing webapp with user-provided plugins](https://github.com/j-blue-arz/image-processing-wasm)
> Image processing webapp with user-provided operators
- Developer can implement and upload image operators in WebAssembly format.
- User can apply operators on an image.

**Reply by nilslice** [reply]
Thanks, @j-blue-arz!

**Comment by nilslice** [comment]
## [antz](https://github.com/hardliner66/antz)

> A small game inspired by AntMe! where you control an ant colony through scripting in order to get more points.

> The game is written in Rust and uses ggez as game engine.

**Reply by nilslice** [reply]
Thanks, @hardliner66!

**Comment by nilslice** [comment]
## [unt-rust-ed](https://github.com/jebradbury39/unt-rust-ed)

> Easy interface for taking in untrusted Rust code and executing it at near-native speed in a lightweight and secure sandbox using WASM

> This library provides a simple api for taking in some untrusted rust code and executing it. It does this by compiling the rust to wasm, and then runs the wasm in a lightweight container (the containerization is mainly provided by Extism). The main benefit of this crate is that it will change the types as needed, so the untrusted code does not necessarily need to be aware that it will be executed as an Extism plugin. It also will take care of building the code into wasm.

### Examples

```rust
#[exported_host_type]
pub struct Inputs {
  pub a: i32,
  pub b: i32,
}

impl Inputs {
  fn new(a: i32, b: i32) -> Self {
    Self { a, b}
  }
}

fn main() {
  let rust_code = "pub fn add(inputs: Inputs) -> i32 {\nreturn inputs.a + inputs.b;\n}";       
                                                                                
  let project = UntrustedRustProject::new(rust_code)
    .with_exported_host_type::<Inputs>();                   
                                                                                
  let compiled_project = project.compile().unwrap();

  let mut container = compiled_project.create_container().unwrap();            
                                                                                
  let outputs: i32 = container.call("add", Inputs::new(10, 2)).unwrap();          
                                                                                
  println!("output: {}", outputs); // prints "12"
}
```

**Reply by nilslice** [reply]
Thanks, @jebradbury39!

**Reply by Zireael07** [reply]
Near native speed is a tall claim to make!

**Comment by nilslice** [comment]
## [brack](https://github.com/brack-lang/brack)

> A bracket-based lightweight markup language that extends commands with WebAssembly

**Reply by nilslice** [reply]
Thanks, @momeemt!

**Reply by srid** [reply]
This looks interesting. Are there any information out there? The repo has no README.

**Reply by Zireael07** [reply]
@srid There seems to be a README now but no examples of syntax it uses

**Reply by momeemt** [reply]
@srid @Zwiterrion
Hello, sorry for the delay in responding. I added a more detailed README. If you need more information and documents about this project, please feel free to ask. Thanks for your interest in Brack!

https://github.com/brack-lang/brack

**Comment by nilslice** [comment]
## [OpenAgents](https://github.com/OpenAgentsInc/openagents)

> An open platform for AI agents, built in public from scratch. 

[Watch this being built!](https://github.com/OpenAgentsInc/openagents/wiki/Video-Series)

![image](https://github.com/extism/extism/assets/7517515/f3f4ecad-47d2-49e7-9335-5b6b3a0946b2)

### How it works
- A user creates an Agent from Nodes.
- Nodes are defined by community developers. They could be:
- API endpoints
- External WASM plugins
- Conditional logic
- Data parsing
- Or most anything compatible with flow-based programming
- Nodes may have an associated fee which is paid to the node creator when the node is used in a workflow
- Agents can be used in our UI and via API
- Users can comment/rate/share Nodes
- Leaderboards show what's popular

**Reply by nilslice** [reply]
Thanks, @atlantispleb!

**Comment by nilslice** [comment]
## [Flowato](https://github.com/graham-hughes-code/flowato-prototype)

> A flow based programming language.

![image](https://github.com/extism/extism/assets/7517515/17b59ae0-98ee-40d3-98ce-51e2be2418fd)

**Reply by nilslice** [reply]
Thanks, @graham-hughes-code!

**Comment by nilslice** [comment]
## [GIF Trick](https://github.com/wymcg/gif_trick)

> A [Matricks](https://github.com/wymcg/matricks) plugin that plays any GIF. By default, this plugin plays catjam, but the plugin can be [easily modified](https://github.com/wymcg/gif_trick?tab=readme-ov-file#change-the-gif) to play any other GIF.

![gif example](https://private-user-images.githubusercontent.com/3410869/281605617-31d2c778-9e17-4f59-9186-e865adf5dc71.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgzNjkzNjksIm5iZiI6MTcwODM2OTA2OSwicGF0aCI6Ii8zNDEwODY5LzI4MTYwNTYxNy0zMWQyYzc3OC05ZTE3LTRmNTktOTE4Ni1lODY1YWRmNWRjNzEuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTlUMTg1NzQ5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2NiMDgyNjdkYTQ2ZTJhYzI5NGI1MGQ3MDc2ODQzMDU4ZmY0MzVjMjA3NTRmMTdmZGMxNmMyZTBlNWJkNTY4MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.udptoI7d_NhO6PzXsp-cYLeWGZ9dEN3hYmhkJYl3VSk)

**Comment by nilslice** [comment]
## [OpenWRT Lua Parser](https://github.com/metaworm/openwrt-luac-parser)

> This is a customized parser for a modified version of luac for openwrt. It can be compiled into WASM and loaded into [metaworm's luadec](https://luadec.metaworm.site/) to decompile openwrt's lua script.

**Reply by nilslice** [reply]
Thanks, @metaworm!

**Comment by nilslice** [comment]
## [Libp2peasy, Lemon Squeezy 🍋✊](https://github.com/DougAnderson444/libp2peasy)

> Ready made [Libp2p](https://libp2p.io/) [Rust Server](https://github.com/libp2p/rust-libp2p/) that connects to the IPFS DHT and gossipsib.

> Design goal is to have all modules separated for easy re-use in other libraries, and yto use a plugin system to extend the functionality of the server as you need it.

**Reply by nilslice** [reply]
Thanks, @DougAnderson444!

**Comment by nilslice** [comment]
## [ZFlow](https://github.com/zflow-dev/zflow)

> Rust implementation of Flow Based Programming Graph specification

This library provides a Rust implementation of a [Flow-Based Programming graphs](https://flow-based.org/) and a runtime for executing the graph. There are two areas covered:

[Graph](https://github.com/darmie/zflow/blob/main/zflow_graph) - the actual graph library
[Journal trait](https://github.com/darmie/zflow/blob/main/zflow_graph/src/journal.rs) - journal system for keeping track of graph changes and undo history
[Graph Runtime](https://github.com/darmie/zflow/blob/main/zflow_runtime) - the process runtime to execute graph components

> See [full example](https://github.com/darmie/zflow/blob/main/zflow_runtime/src/component_test.rs#L396) that demonstrates generation of an HTML code using ZFlow.

**Reply by nilslice** [reply]
Thanks, @darmie!

**Comment by nilslice** [comment]
## [Modsurfer](https://modsurfer.dylibso.com)

![image](https://github.com/extism/extism/assets/7517515/b559621c-1dfa-437f-ae72-52f054d96b20)


Use the CLI: https://github.com/dylibso/modsurfer 
- Extism is used to run [`modsurfer.wasm`](https://modsurfer.dylibso.com/module?hash=0c20c61f67108ebccae1db0be6df7c7d14b2567d5606154278ee390f43e1f408), in particular the `parse_module` function, which extracts information about any wasm module.

**Comment by Zwiterrion** [comment]
# [Wasmo](https://github.com/MAIF/wasmo)

![image](https://github.com/user-attachments/assets/1831b365-33ae-4732-9883-1792cf3cfab5)

> Server and CLI to build WASM plugin using Extism.

**Comment by Zwiterrion** [comment]
# [Wasm4s](https://github.com/MAIF/wasm4s)

> A library to easily run WASM vms from inside your Scala project using Extism

**Comment by chrisdickinson** [comment]
# [Blextism](https://github.com/dylibso/blextism)

-----

> Blender Wasm bindings to Rust using the Python SDK

<div>
    <a href="https://www.loom.com/share/0d73ec07c47644cd8f106260a3d3f61a">
      <p>A walkthrough video:</p>
    </a>
    <a href="https://www.loom.com/share/0d73ec07c47644cd8f106260a3d3f61a">
      <img style="max-width:640px;" src="https://cdn.loom.com/sessions/thumbnails/0d73ec07c47644cd8f106260a3d3f61a-1703011475711-with-play.gif">
    </a>
</div>

**Comment by tsirysndr** [comment]
# [Fluent CI Engine](https://github.com/fluentci-io/fluentci-engine)

Programmable CI/CD with Extism based plugin system

Usage with GitHub Actions :

```yaml
name: Deploy
on:
  push:
    branches:
      - main
jobs:
  deploy:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Setup Fluent CI
        uses: fluentci-io/setup-fluentci@v5
        with:
           wasm: true
           pipeline: cloudflare
           args: deploy
        env:
          CLOUDFLARE_API_TOKEN: ${{ secrets.CF_API_TOKEN }}
          CLOUDFLARE_ACCOUNT_ID: ${{ env.CF_ACCOUNT_ID }} 

```

![Screenshot 2024-04-10 5 56 51 PM](https://github.com/extism/extism/assets/15877106/1d0df081-4bdf-4c35-9a7b-d13869097711)
![Screenshot 2024-04-10 5 59 55 PM](https://github.com/extism/extism/assets/15877106/b6eef992-3c66-4897-ac97-47c4a8fd31b5)

**Comment by aschrijver** [comment]
# [Lemmy](https://join-lemmy.org/)

_Lemmy is a selfhosted social link aggregation and discussion platform. It is completely free and open, and not controlled by any company. This means that there is no advertising, tracking, or secret algorithms. Content is organized into communities, so it is easy to subscribe to topics that you are interested in, and ignore others. Voting is used to bring the most interesting items to the top._

![Screenshot showing a list of posts aggregated from different federated servers.](https://join-lemmy.org/static/assets/images/main_screen_2.webp)

Lemmy, the popular link aggregator for the Fediverse and Free Software decentralized alternative to Reddit, [PR'ed a Proof of Concept](https://github.com/LemmyNet/lemmy/pull/4695) for their extension mechanism. The PoC will be merged only after first plugins become available. Here's what @nutomic wrote on the PR:

> This PR adds a basic plugin hook using [Extism](https://github.com/extism/extism) (webassembly), including a simple example plugin written in Go. Plugins can also be written in many other languages like JS, Rust, .NET or C.
> 
> For now there are only two plugin hooks available, before and after a new post is created through the local api. However we can add all sorts of plugin hooks, eg for different api actions or incoming federated actions. The question is which plugin hooks are actually needed. So if you are interested in writing Lemmy plugins, please comment below with the hooks that you need.
> 
> I dont intend to merge this PR until there is at least a single real-world plugin in development.

**Comment by nilslice** [comment]
# [Oura](https://github.com/txpipe/oura)

Oura is a rust-native implementation of a pipeline that connects to the tip of a Cardano node through a combination of Ouroboros mini-protocol (using either a unix socket or tcp bearer), filters the events that match a particular pattern and then submits a succinct, self-contained payload to pluggable observers called "sinks".

![oura-extism](https://github.com/txpipe/oura/raw/main/assets/demo.svg)

### Filters
- Parse block / tx CBOR
- Split block into txs
- Select Txs by matching rules (address, metadata, policies, etc)
- Enrich tx data with related inputs
- Custom Typescript code (uses Deno)
- Custom WASM plugin
- Rollback buffer with compensating actions

**Reply by nilslice** [reply]
Thanks, @scarmuega!

**Comment by nilslice** [comment]
# [1Password Go SDK](https://github.com/1Password/onepassword-sdk-go)

The 1Password Go SDK offers programmatic access to your secrets in 1Password with Go. During the beta, you can create, retrieve, update, and delete items and resolve secret references.

**Comment by b1nhack** [comment]
# [PumpBin](https://github.com/pumpbin/pumpbin)

🎃 PumpBin is an Implant Generation Platform.

This is a cybersecurity tool that can help cybersecurity teams execute offensive missions more effectively.

![](https://github.com/iced-rs/iced/assets/120295547/7c5107a1-40dd-4735-b3bf-62e68aaa67b3)

**Reply by mhmd-azeez** [reply]
Very cool!

**Comment by aschrijver** [comment]
## [Tau](https://github.com/taubyte/tau)

> Open Source Git-Native CDN PaaS. An alternative to: Vercel, Netlify, Cloudflare, ...
>
> Tau is a framework for building low maintenance & highly scalable cloud computing platforms that software developers will love!
> `tau` is a single binary with no external dependencies except standard system libraries. On top of that, it requires minimal configuration.

![Tau banner image taken from their repository README presenting Tau as 'Cloud development kit'.](https://github.com/user-attachments/assets/01b3daa9-d152-478e-bda4-f4e415d7573f)

Tau [on Hacker News](https://news.ycombinator.com/item?id=40946033)

**Comment by nilslice** [comment]
# [MatrixOne](https://github.com/matrixorigin/matrixone)

> MatrixOne is a hyper-converged cloud & edge native distributed database with a structure that separates storage, computation, and transactions to form a consolidated HSTAP data engine. This engine enables a single database system to accommodate diverse business loads such as OLTP, OLAP, and stream computing. It also supports deployment and utilization across public, private, and edge clouds, ensuring compatibility with diverse infrastructures.

![image](https://github.com/user-attachments/assets/d6b4fca3-2128-4806-9add-8930de1b96ff)

**Comment by nilslice** [comment]
# [Assembllm](https://github.com/bradyjoslin/assembllm)

![image](https://github.com/user-attachments/assets/fb450034-7c22-4631-bf26-5d94ea8ab092)

A versatile CLI tool designed to combine multiple Large Language Models (LLMs) using a flexible task-based system. With a unique WebAssembly-based plugin architecture, it supports seamless integration of various AI models and custom scripts.

### Key Features:

- **Flexible Scripting**: Use pre- and post-scripts for data transformation and integration.
- **Prompt Iteration**: Provide an array of prompts that execute sequentially.
- **Task Chaining**: Chaining multiple tasks into workflows, where outputs of each task feed into the next.
- **Workflow Chaining**: Workflows can call other workflows, allowing you to break down complex operations into smaller, reusable chunks and dynamically link them together.
- **Function / Tool Calling**: Convert unstructured prompts to structured data with OpenAI, Anthropic, and Cloudflare.
- **Multi-Model Support**: Available plugins for [OpenAI](https://platform.openai.com/docs/guides/text-generation/chat-completions-api), [Perplexity](https://docs.perplexity.ai/), [Cloudflare AI](https://developers.cloudflare.com/workers-ai/models/#text-generation), and [Anthropic](https://docs.anthropic.com/en/docs/intro-to-claude).
- **Plugin Architecture**: Extensible via WebAssembly plugins written in a variety of languages, including JavaScript, Rust, Go, and C#, using [Extism](https://extism.org/).

**Comment by aschrijver** [comment]
# [Hyper MCP](https://github.com/tuananh/hyper-mcp)

>`hyper-mcp` is a powerful MCP (Model Context Protocol) server that leverages WebAssembly plugins to extend its capabilities. At its core, it's a single, extensible MCP server that you can enhance with plugins written in any WebAssembly-compatible programming language. Think of it as a modular toolkit where each plugin adds new functionality without the overhead of running multiple MCP servers.
>
>Whether you're using Claude Desktop, Cursor IDE, or any other MCP-compatible application, `hyper-mcp` seamlessly integrates with your workflow while using standard OCI registries for plugin distribution.

**Comment by aschrijver** [comment]
# [Decent Auth](https://github.com/lastlogin-net/decent-auth)

A simple, single authentication library for many languages. Created by [LastLogin](https://lastlogin.net/).

Features:

- Simple
- Compiled to WebAssembly (using [Extism](https://extism.org/)) so it can run in many languages (currently supports Rust, Node.js, and Go. New implementations can be added in <1000 lines of code).
- Focus on decentralized login protocols (currently supports OpenID Connect, ATProto OAuth, Fediverse OAuth, email, and QR code)

Comment found [on Hacker News](https://news.ycombinator.com/item?id=43482556) from the developer:

> I started using WebAssembly in earnest a few months ago to make a backend auth library that works in several different languages[0]. It's built on Extism[1], which abstracts away some of the interfacing complexity. It's been an awesome experience. Frequently feels like magic.
>
>WASM is in an interesting place. The value has clearly been proved with a pretty minimal core spec. Now there's a big push to implement a much larger API surface for WASI and the Component Model. A lot of people in the community are concerned about this direction, or at least the way it's happening[2].
>
>For my part, I hope WASM doesn't go the way of the rest of web browsers where it gets so complicated that only big tech is capable of making implementations and experimenting.
>
>[0]: https://github.com/lastlogin-net/decent-auth
>[1]: https://extism.org/
>[2]: https://www.assemblyscript.org/standards-objections.html

**Comment by SelfDrivingCarp** [comment]
[AutonomousKoi](https://autonomouskoi.org/) is an open source bot framework primarily focused on DJs who stream on Twitch. It's a modular system where components communicate by passing protocol buffer messages over pub-sub-like channels. New functionality can be added through WASM plugins, facilitated by extism.

Modules extend functionality by communicating with DJ hardware/software, calling external APIs, interacting Twitch chat, and providing visuals through OBS Browser Source overlays. The framework will serve local web content packaged with the plugin and that web content and communicate with the rest of the system via WebSockets.

**Comment by nilslice** [comment]
# [Sup](https://github.com/rubiojr/sup)

## WhatsApp Bot framework and CLI with a pluggable architecture

Handlers: with full access to the operating system services, written in Go

Plugins: Sandboxed WASM modules can be developed in several languages (any language supported by [Extism](https://extism.org/docs/quickstart/plugin-quickstart)).

**Comment by mJehanno** [comment]
# [Boot](https://github.com/bootengine/boot)

Boot is a project bootstrapper. It helps you scaffold any type of project while supporting both command execution **and** templating. The scaffolder is define using a config file format (yaml and json supported at the moment, more to come if the need is there). It relies on extism for most of the thing including command support and templating (which means that if, for example, there is no actual support for nodejs command or pug templating, you can jut add it by creating a plugin)

Doc can be found [here](https://bootengine.github.io/)

It's a side project so I'll try my best to take feedback into account but it will take some time.

**Comment by nilslice** [comment]
# [Helm v4](https://helm.sh)

Helm is the best way to find, share, and use software built for [Kubernetes](https://kubernetes.io/).
  
## Examples
https://github.com/scottrigby/h4-example-plugins

## Docs

https://helm.sh/docs/plugins/

**Comment by syke99** [comment]
**[Natyv](https://github.com/natyv-io)**

Natyv is a desktop development framework that removes the need for a WebView or a bundled browser engine, reducing memory footprints from leaky JS runtimes. The host is written in Zig, while actual logic lives in the WASM module. It comes with its own JSX-like syntax that transpiles to pure gust language code, CSS-like styling system without a CSSOM, and a custom Volar-style LSP layer over guest language LSP. It also addresses WASM32's linear memory growth via a custom refresh protocol that refreshes and automatically reattaches event handlers to keep memory pressure low. Currently in beta, with only a Go guest SDK. But all languages with an Extism PDK will eventually be supported.

Being in beta, there's plenty of possibility that there are edge cases and bugs to be found, so I'll attempt to address any feedback as soon as possible. But being a solo dev, it may take some time for me to be able to address them. But any and all feedback, questions, and comments are welcome!
